lib/web/resources.go

```diff
@@ -293,22 +305,58 @@ func upsertTrustedCluster(ctx context.Context, clt resourcesAPIGetter, content,
 	return ui.NewResourceItem(tc)
 }
 
-// CheckResourceUpsertableByError checks if the resource is upsertable by the state of error with
-// the request http method used.
-func CheckResourceUpsertableByError(err error, httpMethod, resourceName string) error {
-	if err != nil && !trace.IsNotFound(err) {
+// resourceExists returns a NotFound error if the resource does not exist.
+type resourceExists func(context.Context, string) error
+
+// CheckResourceUpsert checks if the resource can be created or updated, depending on the http method.
+func CheckResourceUpsert(ctx context.Context, httpMethod string, params httprouter.Params, payloadResourceName string, exists resourceExists) error {
+	switch httpMethod {
+	case http.MethodPost:
+		return trace.Wrap(checkResourceCreate(ctx, payloadResourceName, exists))
+	case http.MethodPut:
+		resourceName := params.ByName("name")
+		if resourceName == "" {
+			return trace.BadParameter("missing resource name")
+		}
+		return trace.Wrap(checkResourceUpdate(ctx, payloadResourceName, resourceName, exists))
+	default:
+		return trace.NotImplemented("http method %q not expected. this is a bug!", httpMethod)
+	}
+}
+
+// checkResourceCreate checks if the resource can be created, returning nil if it can.
+func checkResourceCreate(ctx context.Context, payloadResourceName string, exists resourceExists) error {
+	// Check if a resource with the name in the payload exists.
+	if err := exists(ctx, payloadResourceName); err != nil {
```

Comment:
The name `exists` suggests to me that the function returns a boolean but that's not the case.

What if we renamed it to `getResource` or `getResourceByName` and ignored the first return value? We could also let go of the nesting to express the purpose of `checkResourceCreate` in a more "linear" manner.

```go
// checkResourceCreate checks if the resource can be created, returning nil if it can.
func checkResourceCreate(ctx context.Context, payloadResourceName string, getResource resourceExists) error {
	// Check if a resource with the name in the payload exists.
	_, err := getResource(ctx, payloadResourceName)

	if trace.IsNotFound(err) {
		// The resource does not exist and can be created.
		return nil
	}

	if err == nil {
		// The resource was found and a new one with the same name cannot be created.
		return trace.AlreadyExists("resource name %q already exists", payloadResourceName)
	}

	return trace.Wrap(err)
}
```

I realize this is a bit unconventional so feel free to drop this suggestion. OTOH what we do here is also "unconventional" as the presence of a specific error _is_ the happy path of this function.

Alternatively, we could also swap the checks so that first we check if error is nil and then the rest of the function can assume that it's present.

<details>
<summary>Check err == nil first</summary>

```go
// checkResourceCreate checks if the resource can be created, returning nil if it can.
func checkResourceCreate(ctx context.Context, payloadResourceName string, getResource resourceExists) error {
	// Check if a resource with the name in the payload exists.
	_, err := getResource(ctx, payloadResourceName)

	if err == nil {
		// The resource was found and a new one with the same name cannot be created.
		return trace.AlreadyExists("resource name %q already exists", payloadResourceName)
	}

	if trace.IsNotFound(err) {
		// The resource does not exist and can be created.
		return nil
	}

	return trace.Wrap(err)
}
```
</details>

Reply:
Thanks for the review and the suggestion! Did something similar in 4e53ac1afc65dc7a306a43a6b376652114160844.